File: big data(3 project)/thema_7/thema_7.1.py
#import necessary libraries
import csv
import pandas as pd
from kmodes.kmodes import KModes
import numpy as np
# Specify the path to your CSV file
csv_file_path1 = 'C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\movies.csv'

# Open the CSV file with specified encoding (e.g., UTF-8)
with open(csv_file_path1, 'r', encoding='utf-8') as file:
    # Create a CSV reader object
    csv_reader = csv.reader(file)

    # Convert the CSV data into a list of lists
    data = [row for row in csv_reader]

# Create a Pandas DataFrame from the list of lists
movies_df = pd.DataFrame(data)
print(movies_df)
csv_file_path2 = 'C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\ratings.csv'
# Open the CSV file with specified encoding (e.g., UTF-8)
with open(csv_file_path2, 'r', encoding='utf-8') as file:
    # Create a CSV reader object
    csv_reader = csv.reader(file)

    # Convert the CSV data into a list of lists
    data = [row for row in csv_reader]

# Create a Pandas DataFrame from the list of lists
ratings_df = pd.DataFrame(data)
print(ratings_df)

# Extract only the category variables (dummy variables) from the movies DataFrame(starting from the 3rd column) 
categories_data = movies_df.iloc[:, 2:]

# Choose the number of clusters (K) using the Elbow method
# Assume that k_values is a list containing values from 2 to 100
sum_of_squared_distances = []
k_values = range(2, 101)
#loop over each k in k_values
for k in k_values:
    #initialize kmodes cluster model
    #init, Specifies the initialization method. 'Huang' is one of the initialization methods available in the KModes algorithm.
    #It is a method that initializes the centroids based on the density of the categorical data.
    #n_init specifies the number of times the K-modes algorithm will be run with different centroid seeds.
    #This is done to mitigate the impact of random initialization, and the best result in terms of clustering cost will be retained.
    #int=5 is the number of times the algorythm will be run with different centroid seeds. This is done to mitigate the effect of the random initialization
    #of centroids at the begining of the algorythm, something which does affect the final outcome, basically it represents a trade-off between computational
    #cost and obtaining a robust solution.
    #the number 5 is chosen since it is not too big(high comutatonal cost) neither too small(loss in robustness of results)
    kmeans = KModes(n_clusters=k, init='Huang', n_init=5, verbose=0)
    #fits the model to the categories data
    kmeans.fit_predict(categories_data)
    #After fitting and predicting, this attribute stores the cost of the clustering, which is the sum of squared distances.
    #The lower the cost, the better the clustering.

    sum_of_squared_distances.append(kmeans.cost_)

# Plot the Elbow graph
import matplotlib.pyplot as plt
#design the plot of the elbow graph
plt.plot(k_values, sum_of_squared_distances, 'bx-')
plt.xlabel('Number of Clusters (K)')
plt.ylabel('Sum of Squared Distances')
plt.title('Elbow Method For Optimal K')
plt.show()

# Choose the optimal number of clusters (K) based on the Elbow method
#This funcion finds the index where the first derivative of the sum of squared distances is minimized
optimal_k = np.argmin(np.diff(sum_of_squared_distances)) + 2  # Assuming K starts from 2
print(f"Optimal number of clusters (K): {optimal_k}")
print("doone")
#re-run the k-clustering model with the optimal n of clsters
kmeans = KModes(n_clusters=optimal_k, init='Huang', n_init=5, verbose=0)
# fits k-model to categories and predict cluster assignment of each data point
cluster_labels = kmeans.fit_predict(categories_data)
#add column to movies_df with the cluster id for each data point 
movies_df['clusterId'] = cluster_labels
# Calculate the average rating for each movie
average_ratings = ratings_df.groupby('movieId')['rating'].mean()
#merge the two df based on the unique "movieId" left_on mean that the "movieId" column of the movies_df will be used as key for merging, while right_index=True that
#index of the average_ratings will be used as key for merging, how='left', refers to the type of merge (left joint)
#The resulting merged df will contain all columns of the movies_df + the average rating of each movie, if a movie doe not have a corresponding entry in average ratings
#then the value of the column will be NaN 
movies_df = pd.merge(movies_df, average_ratings, left_on='movieId', right_index=True, how='left')

#Specifie the user for whom movie recommendations are being generated
user_id = 198
# extract ratings from the ratings_df for the specific user 
user_ratings = ratings_df[ratings_df['userId'] == user_id]

# Merge user_ratings_df and movies_df on movieId and clusterId
user_ratings = pd.merge(user_ratings, movies_df[['movieId', 'clusterId']], on='movieId', how='left')

# Group user ratings by clusterId and calculate the mean rating for each cluster
cluster_avg_ratings = user_ratings.groupby('clusterId')['rating'].mean()

# Display the average ratings for each cluster
print("\nAverage Ratings for Each Cluster:")
for cluster_id, avg_rating in cluster_avg_ratings.iteritems():
    print(f"Cluster {cluster_id}: {avg_rating:.2f}")#2f is for the numer of digits displayed

# Filter clusters with average rating >= 3.5(.index retrievs the index values(clusterId's))
high_rated_clusters = cluster_avg_ratings[cluster_avg_ratings >= 3.5].index

# Check if there are clusters with average rating >= 3.5
if not high_rated_clusters.empty:
    # For each high-rated cluster, find top 2 movies user hasn't seen
    #empty df 

    recommendations = pd.DataFrame()
    for cluster_id in high_rated_clusters:
        #keep only movies that belong to the current cluster
        cluster_movies = movies_df[movies_df['clusterId'] == cluster_id]
        #keep only the unrated movies using '~'
        unrated_movies = cluster_movies[~cluster_movies['movieId'].isin(user_ratings['movieId'])]
        
        # Check if there are unrated movies in the cluster
        if not unrated_movies.empty:
            #if unrated movie in cluster select the top 2 (with the highest rating)
            top_movies = unrated_movies.nlargest(2, 'rating')
            #put the in the recommndations df
            recommendations = recommendations.append(top_movies)

    # Check if there are recommendations
    if not recommendations.empty:
        # Display recommendations
        print("\nMovie Recommendations:")
        #for each row in recommendations df 
        for _, movie in recommendations.iterrows():
            print(f"Movie Title: {movie['title']}, Average Rating: {movie['rating']:.2f}")
    else:
        print("\nSorry, no new movies to recommend from high-rated clusters!")

else:
    print("\nSorry, no high-rated clusters found for recommendations!")

File: big data(3 project)/thema_7/thema_7_1.R
# Remove all objects from the workspace
rm(list = ls())

# Set the working directory
setwd("C://Users//kwstasbenek//Desktop/giannakopoulo_8//Tsagkarakis_3h_ergasia")
#libraries/packages
install.packages(c("dplyr", "tidyr", "cluster", "ggplot2"))
library(dplyr)
library(tidyr)
library(cluster)
library(ggplot2)

# Read movies and ratings data
movies <- read.csv('C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\movies.csv', header = TRUE)
ratings <- read.csv('C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\ratings.csv', header = TRUE)

# Extract category variables (dummy variables) from the movies DataFrame
categories_data <- movies[, 3:ncol(movies)]

# Choose the number of clusters (K) using the Elbow method
# Assume that k_values is a vector containing values from 2 to 100
sum_of_squared_distances <- numeric(length = 0)
k_values <- 2:100

for (k in k_values) {
  # Initialize k-modes cluster model
  #iteration=5 for the same reason as metioned in the python script
  kmeans_result <- KMeans(categories_data, centers = k, algorithm = "Hartigan-Wong", iter.max = 5, nstart = 5)
  
  # After fitting and predicting, store the cost of clustering (sum of squared distances)
  #The withinss attribute represents the sum of squared distances from each point in a cluster to the center of that cluster. 
  #Essentially, it is a measure of how tight or compact the clusters are.
  #Then, we append the total sum of squared distances for the current value of k (number of clusters) to the vector
  sum_of_squared_distances <- c(sum_of_squared_distances, sum(kmeans_result$withinss))
}

# Plot the Elbow graph
plot(k_values, sum_of_squared_distances, type = 'b', xlab = 'Number of Clusters (K)', ylab = 'Sum of Squared Distances', main = 'Elbow Method For Optimal K')

# Choose the optimal number of clusters (K) based on the Elbow method
# Find the index where the first derivative of the sum of squared distances is minimized
#The diff(sum_of_squared_distances) calculates the differences between consecutive elements in the vector sum_of_squared_distances. 
#The which.min() function is then applied to find the index where these differences are minimized.
optimal_k <- which.min(diff(sum_of_squared_distances)) + 1 #assuming k starts from 2


cat("Optimal number of clusters (K):", optimal_k, "\n")

# Re-run the k-clustering model with the optimal number of clusters
kmeans_result <- kmeans(categories_data, centers = optimal_k, algorithm = "Hartigan-Wong", iter.max = 5, nstart = 5)
cluster_labels <- kmeans_result$cluster

# Add the 'clusterId' column to the movies data frame
movies$clusterId <- cluster_labels

# Calculate the average rating for each movie
#FUN = mean: Specifies the function to be applied to each subset of data. 
#Here, we are using the mean function to calculate the average rating for each movie.
average_ratings <- aggregate(rating ~ movieId, data = ratings, FUN = mean)

# Merge the two data frames based on the unique "movieId"
#all.x specifies that all the rows from the left data frame (movies in this case) should be included in the merged data frame.
movies <- merge(movies, average_ratings, by = 'movieId', all.x = TRUE)

# Specify the user for whom movie recommendations are being generated
user_id <- 198

# Extract ratings from the ratings data frame for the specific user
user_ratings <- subset(ratings, userId == user_id)

# Merge user_ratings_df and movies_df on movieId and clusterId
#cbind
user_ratings <- merge(user_ratings, movies[, c('movieId', 'clusterId')], by = 'movieId', all.x = TRUE)

# Group user ratings by clusterId and calculate the mean rating for each cluster
#aggregate function is used to to group the 'user_ratings' data frame by the 'clusterId' column and calculate the mean rating for each cluster.
cluster_avg_ratings <- aggregate(rating ~ clusterId, data = user_ratings, FUN = mean)

# Display the average ratings for each cluster
cat("\nAverage Ratings for Each Cluster:\n")
print(cluster_avg_ratings)

# Filter clusters with average rating >= 3.5
#and keep their clusterId
high_rated_clusters <- subset(cluster_avg_ratings, rating >= 3.5)$clusterId

# Check if there are clusters with average rating >= 3.5
#check if not empty
if (length(high_rated_clusters) > 0) {
  # For each high-rated cluster, find top 2 movies user hasn't seen
  recommendations <- data.frame()
  
  for (cluster_id in high_rated_clusters) {
    # Keep only movies that belong to the current cluster and subset them
    
    cluster_movies <- subset(movies, clusterId == cluster_id)
    
    # Keep only the unrated movies using '~'
    #!(movieId %in% user_ratings$movieId): This part is a logical condition. 
    #It checks for each movie in cluster_movies whether its movieId is NOT in the vector of movie IDs that the user has rated (user_ratings$movieId).
    #Then we subset based on this logical condition
    unrated_movies <- subset(cluster_movies, !(movieId %in% user_ratings$movieId))
    
    # Check if there are unrated movies in the cluster
    if (nrow(unrated_movies) > 0) {
      # If unrated movies in the cluster, select the top 2 (with the highest rating)
      #we order the unrated _movies df in a decreasing order and only keep the first 2 movies
      top_movies <- unrated_movies[order(unrated_movies$rating, decreasing = TRUE), ][1:2, ]
      
      # Put them in the recommendations data frame
      recommendations <- rbind(recommendations, top_movies)
    }
  }
  
  # Check if there are recommendations
  if (nrow(recommendations) > 0) {
    # Display recommendations
    cat("\nMovie Recommendations:\n")
    print(recommendations)
  } else {
    cat("\nSorry, no new movies to recommend from high-rated clusters!\n")
  }
  
} else {
  cat("\nSorry, no high-rated clusters found for recommendations!\n")
}
